infor-design/enterprise
# 962/1038 - Add "Node-Sass" wrapper script / Fix incorrect SCSS source maps

**Explain the _details_ for making this change. What existing problem does the pull request solve?**
This PR Improves the CSS building system in many ways:
- Adds a custom harness for Node Sass (#1038)
- Adds a system to the IDS Custom Builder for inlining the copyright banner during CSS bundling (also fixes source map problems)
- Added more efficient minification for both JS and CSS (minification changes on the CSS side also help #962)
- Makes the IDS Custom Builder the default bundling system (removed the Grunt task)
- Modifies the demoapp to serve the `/src` folder as a static asset (helps make the links inside CSS in the debugger actually work)
- General cleanup work on the IDS Custom Builder

**Related github/jira issue (required)**:
- Closes #962 
- Partially covers #1038 (Node-Sass wrapper has been implemented)

**Steps necessary to review your pull request (required)**:

_Test a standard build:_
- Run `npx grunt`.  The entire task should complete
- Open the `dist/css/` folder.  Observe that each theme should have a standard CSS, minified CSS, and a MAP file for the standard CSS.  The two CSS files should both have an inlined comment banner containing the standard copyright/build info/date/time. 
- Open this `dist/js` folder.  Observe that `sohoxi.js`, `sohoxi.min.js` exist each with a corresponding `.map` file.

_Test dev tools debugging:_
- Run `npm start`
- Open a browser and navigate to http://localhost:4000/components/toolbar-flex/example-index
- Open a dev tools console to the Elements panel.  Use the "Inspect Element" cursor on any of the `.toolbar-section` elements.
- In the "Styles" panel, ensure that the CSS rules present all correctly link back to the source code file and line specified:

![screen shot 2018-11-29 at 4 39 09 pm](https://user-images.githubusercontent.com/3249601/49253786-71685480-f3f5-11e8-9a77-aa8af19f15dd.png)
![screen shot 2018-11-29 at 4 39 17 pm](https://user-images.githubusercontent.com/3249601/49253787-71685480-f3f5-11e8-8028-30460c7ecb77.png)

_Test the watch task:_
- Run `npm run quickstart` to run the demoapp
- Run `npx grunt watch` to run the watch task
- Change a CSS file in the `/src` directory by adding a junk CSS rule with multiple lines, for example:
```scss
.new-rule {
  .cool-rule {
    .nifty-rule {
      .excellent-rule {
        display: none;
      }
    }
  }
}
```
- The CSS Build should run, and new CSS files and corresponding map files should be generated.
- Re-check the Elements panel and inspect the same elements.  The locations of the CSS rules should be correct when clicking the links in the debugger.
